habitat-lab/habitat/gui/text_drawer.py

```diff
@@ -0,0 +1,119 @@
+
+        # Load a TrueTypeFont plugin and open the font file
+        self._display_font = text.FontManager().load_and_instantiate(
+            "TrueTypeFont"
+        )
+        self._display_font.open_file(
+            os.path.join(os.path.dirname(__file__), relative_path_to_font),
+            13,
+        )
+        self._display_font_size = display_font_size
+        self._max_display_text_chars = max_display_text_chars
+
+        # Glyphs we need to render everything
+        self._glyph_cache = text.GlyphCache(mn.Vector2i(256))
+        self._display_font.fill_glyph_cache(
+            self._glyph_cache,
+            string.ascii_lowercase
+            + string.ascii_uppercase
+            + string.digits
+            + ":-_+,.! %µ",
+        )
+        self._shader = shaders.VectorGL2D()
+        self._window_text = text.Renderer2D(
+            self._display_font,
+            self._glyph_cache,
+            display_font_size,
+            text.Alignment.TOP_LEFT,
+        )
+
+        # make magnum text background transparent
+        mn.gl.Renderer.enable(mn.gl.Renderer.Feature.BLENDING)
+        mn.gl.Renderer.set_blend_function(
+            mn.gl.Renderer.BlendFunction.ONE,
+            mn.gl.Renderer.BlendFunction.ONE_MINUS_SOURCE_ALPHA,
+        )
+        mn.gl.Renderer.set_blend_equation(
+            mn.gl.Renderer.BlendEquation.ADD, mn.gl.Renderer.BlendEquation.ADD
+        )
+
+    def add_text(
+        self,
+        text_to_add,
+        alignment: TextOnScreenAlignment = TextOnScreenAlignment.TOP_LEFT,
+        text_delta_x: int = 0,
+        text_delta_y: int = 0,
+    ):
+        """
+        Adds `text_to_add` and corresponding window text transform to `self._text_transform_pairs`.
+            :param text_to_add: text to be added to `self._text_transform_pairs` for drawing
+            :param alignment: window text anchor of type TextOnScreenAlignment, defaults to top left corner
+            :param text_delta_x: pixels delta to move/adjust window text anchor along X axis,
+            :param text_delta_y: pixels delta to move/adjust window text anchor along Y axis,
+        """
+
+        # text object transform in window space is Projection matrix times Translation Matrix
+        # put text in top left of window
+        align_y, align_x = alignment.value
+        window_text_transform = mn.Matrix3.projection(
+            self._framebuffer_size
+        ) @ mn.Matrix3.translation(
+            mn.Vector2(self._framebuffer_size) * mn.Vector2(align_x, align_y)
+            + mn.Vector2(text_delta_x, text_delta_y)
+        )
+        self._text_transform_pairs.append((text_to_add, window_text_transform))
+
+    def draw_text(self):
+        """Draws collected text on the screen"""
+        for text_to_draw, transform in self._text_transform_pairs:
+            self._shader.bind_vector_texture(self._glyph_cache.texture)
+            self._window_text.reserve(self._max_display_text_chars)
```

Comment:
@mosra what's the right thing to do here? Do one `reserve(max_display_text_chars)` at init time? Or maybe `reserve(len(text_to_draw))` every time here?

Can we also add a runtime check that we're not going over `max_display_text_chars`?

Reply:
One reserve at init time, yes. The mapping from characters to glyphs isn't always 1:1 (depends on the font) so in general it's not as simple as `len()`.

That being said, the `Text` APIs could use a UX refresh, and I have some updates scheduled there as part of my work on other projects.

Reply:
Er, I mean, the code as is here is not ideal perf-wise because it overwrites a single mesh several times over. But that's what you get with the current API, it wasn't made with such "immediate-mode use" in mind. Instead one was supposed to have several dedicated `text.Renderer2D` instances that get updated/rerendered if they actually need an update (for example it probably wouldn't be needed to refresh the values every frame as such rapid changes would be impossible to spot by the human eye anyway).

But as I said, I have some updates scheduled, and making this use case better-supported (such as being able to efficiently fill a larger mesh from smaller text pieces) is among them.